external/radiation/radiation/preprocessing.py

```diff
@@ -0,0 +1,257 @@
+    state_names.extend(tracer_names)
+    states = xr.Dataset({name: state[name] for name in state_names})
+    stacked = _stack(states, unstacked_dim)
+    delp = stacked["pressure_thickness_of_atmospheric_layer"]
+    pi = pressure_at_interface(delp, dim_center=unstacked_dim, dim_outer="z_interface")
+    pm = pressure_at_midpoint_log(delp, dim=unstacked_dim)
+    exn = (pm / P_REF) ** (_RDGAS / _SPECIFIC_HEAT_CONST_PRESSURE)  # exner pressure
+    temperature = stacked["air_temperature"]
+
+    tracer_array = np.zeros(
+        (stacked.sizes["column"], stacked.sizes[unstacked_dim], len(tracer_names))
+    )
+    for tracer_name, index in tracer_inds.items():
+        tracer_array[:, :, index - 1] = stacked[tracer_name].values
+
+    _statein = {
+        "prsi": pi.values,
+        "prsl": pm.values,
+        "tgrs": temperature.values,
+        "prslk": exn.values,
+        "qgrs": tracer_array,
+    }
+    if ivflip == 1:
+        _statein = {name: np.flip(_statein[name], axis=1) for name in _statein}
+    return _statein
+
+
+def _stack(
+    ds: xr.Dataset, unstacked_dim: str = "z", sample_dim: str = "column"
+) -> xr.Dataset:
+    stack_dims = [dim for dim in ds.dims if dim != unstacked_dim]
+    stacked = ds.stack({sample_dim: stack_dims})
+    if unstacked_dim in ds.dims:
+        stacked = stacked.transpose(sample_dim, unstacked_dim)
+    return stacked
+
+
+def unstack(
+    data: Mapping[str, np.ndarray],
+    coords: xr.DataArray,
+    sample_dim: str = "column",
+    unstacked_dim: str = "z",
+) -> Diagnostics:
+    out: Diagnostics = {}
+    for name, arr in data.items():
+        if arr.ndim == 1:
+            da = xr.DataArray(arr, dims=[sample_dim], coords={sample_dim: coords})
+        elif arr.ndim == 2:
+            da = xr.DataArray(
+                arr, dims=[sample_dim, unstacked_dim], coords={sample_dim: coords}
+            )
+        out[name] = da.unstack(dim=sample_dim)
+    return out
+
+
+def grid(state: State) -> Tuple[Mapping[str, np.ndarray], xr.DataArray]:
+    grid_names = ["longitude", "latitude"]
+    stacked_grid = _stack(
+        xr.Dataset({name: state[name] for name in grid_names}), sample_dim="column"
+    )
+    coords = stacked_grid.column
+    lon_rad = stacked_grid["longitude"].values
+    lat_rad = stacked_grid["latitude"].values
+    sinlat = np.sin(lat_rad)
+    coslat = np.cos(lat_rad)
+    return (
+        {"xlon": lon_rad, "xlat": lat_rad, "sinlat": sinlat, "coslat": coslat},
+        coords,
+    )
+
+
+def sfcprop(
+    state: State, sfc_var_mapping: Mapping[str, str] = SFC_NAMES_IN_MAPPING
+) -> Mapping[str, np.ndarray]:
+    sfc = xr.Dataset({v: state[k] for k, v in list(sfc_var_mapping.items())})
+    # we only want the first of "hprime" variables
+    sfc["hprime"] = sfc["hprime"].isel({"orographic_variable": 0})
+    stacked_sfc = _stack(sfc)
+    return {name: stacked_sfc[name].values for name in stacked_sfc.data_vars}
+
+
+def postprocess_out(
```

Comment:
I'm not sure what is meant by flattening, post_processing etc. Maybe just flatten the call stack here and inline it in  the calling context. Do we really parsing/interpretation to do this? I think being verbose/explicit here would be advantageous.

Reply:
this is optional...maybe for a future pr.

Reply:
The radiation driver returns a tortuously nested set of dicts. A future refactor could clean up its API and eliminate the need for this.